# Practical Application III: Comparing Classifiers

**Overview**: In this practical application, your goal is to compare the performance of the classifiers we encountered in this section, namely K Nearest Neighbor, Logistic Regression, Decision Trees, and Support Vector Machines.  We will utilize a dataset related to marketing bank products over the telephone.  



### Getting Started

Our dataset comes from the UCI Machine Learning repository [link](https://archive.ics.uci.edu/ml/datasets/bank+marketing).  The data is from a Portugese banking institution and is a collection of the results of multiple marketing campaigns.  We will make use of the article accompanying the dataset [here](CRISP-DM-BANK.pdf) for more information on the data and features.



### Problem 1: Understanding the Data

To gain a better understanding of the data, please read the information provided in the UCI link above, and examine the **Materials and Methods** section of the paper.  How many marketing campaigns does this data represent?

### Problem 2: Read in the Data

Use pandas to read in the dataset `bank-additional-full.csv` and assign to a meaningful variable name.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('/Users/browngazelle/Python Files (Doug)/Data/bank-additional-full.csv', sep = ';')

In [3]:
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

### Problem 3: Understanding the Features


Examine the data description below, and determine if any of the features are missing values or need to be coerced to a different data type.


```
Input variables:
# bank client data:
1 - age (numeric)
2 - job : type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')
3 - marital : marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)
4 - education (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')
5 - default: has credit in default? (categorical: 'no','yes','unknown')
6 - housing: has housing loan? (categorical: 'no','yes','unknown')
7 - loan: has personal loan? (categorical: 'no','yes','unknown')
# related with the last contact of the current campaign:
8 - contact: contact communication type (categorical: 'cellular','telephone')
9 - month: last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')
10 - day_of_week: last contact day of the week (categorical: 'mon','tue','wed','thu','fri')
11 - duration: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no'). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.
# other attributes:
12 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
13 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)
14 - previous: number of contacts performed before this campaign and for this client (numeric)
15 - poutcome: outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')
# social and economic context attributes
16 - emp.var.rate: employment variation rate - quarterly indicator (numeric)
17 - cons.price.idx: consumer price index - monthly indicator (numeric)
18 - cons.conf.idx: consumer confidence index - monthly indicator (numeric)
19 - euribor3m: euribor 3 month rate - daily indicator (numeric)
20 - nr.employed: number of employees - quarterly indicator (numeric)

Output variable (desired target):
21 - y - has the client subscribed a term deposit? (binary: 'yes','no')
```



In [4]:
# Check for missing values and data types
missing_values = df.isnull().sum()
data_types = df.dtypes

missing_values, data_types

(age               0
 job               0
 marital           0
 education         0
 default           0
 housing           0
 loan              0
 contact           0
 month             0
 day_of_week       0
 duration          0
 campaign          0
 pdays             0
 previous          0
 poutcome          0
 emp.var.rate      0
 cons.price.idx    0
 cons.conf.idx     0
 euribor3m         0
 nr.employed       0
 y                 0
 dtype: int64,
 age                 int64
 job                object
 marital            object
 education          object
 default            object
 housing            object
 loan               object
 contact            object
 month              object
 day_of_week        object
 duration            int64
 campaign            int64
 pdays               int64
 previous            int64
 poutcome           object
 emp.var.rate      float64
 cons.price.idx    float64
 cons.conf.idx     float64
 euribor3m         float64
 nr.employed       float64
 y  

The dataset does not have any missing values for any of the features, as all columns have zero missing entries.

However, all categorical columns (job, marital, education, default, housing, loan, contact, month, day_of_week, poutcome, and y) are represented as objects, which is expected for categorical data. These may need to be encoded depending on the type of model or analysis you plan to perform.

### Problem 4: Understanding the Task

After examining the description and data, your goal now is to clearly state the *Business Objective* of the task.  State the objective below.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

A "term deposit", the client agrees to keep a certain amount of money deposited in the account for a fixed period, during which the bank pays interest. During this period, the money cannot be withdrawn without penalty. Once the term ends, the client can withdraw the funds or renew the term deposit, often at a higher interest rate than a standard savings account.

In the context of the dataset, the question is asking if the client will ultimately agreed to subscribe to this financial product, eg cam we PREDICT the output variable y with a binary response (yes or no).

Excluding "duration" as it is determistic and cannot be done before the call

### Problem 5: Engineering Features

Now that you understand your business objective, we will build a basic model to get started.  Before we can do this, we must work to encode the data.  Using just the bank information features, prepare the features and target column for modeling with appropriate encoding and transformations.

In [8]:
# Drop the 'duration' column
data_dropped = df.drop(columns=['duration'])

# Perform one-hot encoding on the categorical columns
categorical_columns = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome', 'y']
data_encoded = pd.get_dummies(data_dropped, columns=categorical_columns, drop_first=True)

# Display the new dataframe
print(data_encoded.head())


   age  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
0   56         1    999         0           1.1          93.994   
1   57         1    999         0           1.1          93.994   
2   37         1    999         0           1.1          93.994   
3   40         1    999         0           1.1          93.994   
4   56         1    999         0           1.1          93.994   

   cons.conf.idx  euribor3m  nr.employed  job_blue-collar  ...  month_nov  \
0          -36.4      4.857       5191.0            False  ...      False   
1          -36.4      4.857       5191.0            False  ...      False   
2          -36.4      4.857       5191.0            False  ...      False   
3          -36.4      4.857       5191.0            False  ...      False   
4          -36.4      4.857       5191.0            False  ...      False   

   month_oct  month_sep  day_of_week_mon  day_of_week_thu  day_of_week_tue  \
0      False      False             True            Fals

### Problem 6: Train/Test Split

With your data prepared, split it into a train and test set.

In [9]:
from sklearn.model_selection import train_test_split

# Assuming 'y_yes' is the target variable
X = data_encoded.drop(columns=['y_yes'])
y = data_encoded['y_yes']

# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train.shape, X_test.shape, y_train.shape, y_test.shape


((32950, 52), (8238, 52), (32950,), (8238,))

### Problem 7: A Baseline Model

Before we build our first model, we want to establish a baseline.  What is the baseline performance that our classifier should aim to beat?

In [10]:
# Calculate the baseline by finding the most frequent class in the training target
baseline_class = y_train.mode()[0]  # Most frequent class (0 or 1)

# Calculate baseline accuracy by assuming that we predict the most frequent class for all samples
baseline_accuracy = (y_train == baseline_class).mean()

baseline_accuracy


0.887556904400607

### Problem 8: A Simple Model

Use Logistic Regression to build a basic model on your data.  

In [11]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Initialize the logistic regression model
log_reg_model = LogisticRegression(max_iter=1000)

# Train the model using the training data
log_reg_model.fit(X_train, y_train)

# Predict on the test set
y_pred = log_reg_model.predict(X_test)

# Calculate the accuracy of the logistic regression model
log_reg_accuracy = accuracy_score(y_test, y_pred)

log_reg_accuracy


/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.8969410050983249

### Problem 9: Score the Model

What is the accuracy of your model?

The accuracy_score is about 89.69%, beating the benchmark by ~1%

### Problem 10: Model Comparisons

Now, we aim to compare the performance of the Logistic Regression model to our KNN algorithm, Decision Tree, and SVM models.  Using the default settings for each of the models, fit and score each.  Also, be sure to compare the fit time of each of the models.  Present your findings in a `DataFrame` similar to that below:

| Model | Train Time | Train Accuracy | Test Accuracy |
| ----- | ---------- | -------------  | -----------   |
|     |    |.     |.     |

In [13]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from time import time

# Initialize models
models = {
    'Logistic Regression': LogisticRegression(max_iter=1000),
    'K-Nearest Neighbors': KNeighborsClassifier(),
    'Decision Tree': DecisionTreeClassifier(),
    'Support Vector Machine': SVC()
}

# Dictionary to store results
results = {'Model': [], 'Train Time': [], 'Train Accuracy': [], 'Test Accuracy': []}

# Train and evaluate each model
for model_name, model in models.items():
    start_time = time()
    
    # Train the model
    model.fit(X_train, y_train)
    train_time = time() - start_time
    
    # Predict on train and test sets
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)
    
    # Calculate accuracies
    train_accuracy = accuracy_score(y_train, y_train_pred)
    test_accuracy = accuracy_score(y_test, y_test_pred)
    
    # Store results
    results['Model'].append(model_name)
    results['Train Time'].append(train_time)
    results['Train Accuracy'].append(train_accuracy)
    results['Test Accuracy'].append(test_accuracy)

# Convert results into a DataFrame for presentation
results_df = pd.DataFrame(results)

# Display the results
print(results_df)

/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


                    Model  Train Time  Train Accuracy  Test Accuracy
0     Logistic Regression    7.764509        0.901184       0.896941
1     K-Nearest Neighbors    0.053589        0.914719       0.888322
2           Decision Tree    0.543135        0.995357       0.838553
3  Support Vector Machine   34.345328        0.898209       0.894756


Create a READ.ME file for describing the data, the task, the decisions made and the output.  

Update metrics:  
                    Model  Train Accuracy  Test Accuracy  F1-Score   AUC-ROC
0     Logistic Regression        0.901335       0.897062  0.319422  0.780002
1     K-Nearest Neighbors        0.912504       0.889900  0.368824  0.714865
2  Support Vector Machine        0.903672       0.897427  0.319098  0.699183

Per scale and hyperparmeter activities:

                   Model  Train Time  Train Accuracy  Test Accuracy
0     Logistic Regression    7.764509        0.901184       0.896941
1     K-Nearest Neighbors    0.053589        0.914719       0.888322
2           Decision Tree    0.543135        0.995357       0.838553
3  Support Vector Machine   34.345328        0.898209       0.894756

### Problem 11: Improving the Model

Now that we have some basic models on the board, we want to try to improve these.  Below, we list a few things to explore in this pursuit.

- More feature engineering and exploration.  For example, should we keep the gender feature?  Why or why not?
- Hyperparameter tuning and grid search.  All of our models have additional hyperparameters to tune and explore.  For example the number of neighbors in KNN or the maximum depth of a Decision Tree.  
- Adjust your performance metric

I am going to focus on these.

1. Performing hyperparameter tuning with grid search for one of the models (e.g., KNN).
2. Exploring feature scaling or engineering for models sensitive to feature ranges.
3. Switching metrics to something like F1-score or AUC-ROC, depending on the goals.

Though gender likely doesn't have strong coorelation, it was not in the orginal data so will ignore that sugestion. 

I am going to start with KNN given the efficiency of it run time and train and test accuracy.

In [14]:
from sklearn.model_selection import GridSearchCV

# Let's start hyperparameter tuning with K-Nearest Neighbors as an example

# Define the hyperparameter grid for KNN
param_grid_knn = {
    'n_neighbors': [3, 5, 7, 9, 11],  # Number of neighbors to test
    'p': [1, 2]  # 1 for Manhattan distance, 2 for Euclidean distance
}

# Initialize KNN model
knn_model = KNeighborsClassifier()

# Initialize GridSearchCV for KNN
grid_search_knn = GridSearchCV(knn_model, param_grid_knn, cv=5, scoring='accuracy', n_jobs=-1)

# Perform the grid search on the training data
grid_search_knn.fit(X_train, y_train)

# Get the best parameters and best score
best_params_knn = grid_search_knn.best_params_
best_score_knn = grid_search_knn.best_score_

best_params_knn, best_score_knn


({'n_neighbors': 11, 'p': 1}, 0.8944461305007587)

Represents a slight increase. ~1%

## Hyperparamter and scaling the data
Standardization (for KNN, Logistic Regression, and SVM).
Log Transformation for skewed features (like pdays) to normalize their distributions.
Polynomial Features for models like Logistic Regression or SVM, especially if you suspect non-linear relationships between the features and target.

Standardization (Z-score Normalization)
Why: Standardization works well for distance-based models like KNN and SVM, and can also improve the performance of Logistic Regression. Since these models are sensitive to the magnitude of the features, standardizing the numeric features to have a mean of 0 and a standard deviation of 1 helps ensure that each feature contributes equally to the model.
Features to Standardize:
age
campaign
pdays
previous
emp.var.rate, cons.price.idx, cons.conf.idx, euribor3m, and nr.employed


In [15]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


Log Transformation for Skewed Features
Why: Some features like pdays might be highly skewed (many clients were not contacted in previous campaigns, represented by a value like 999). A log transformation can compress the range of skewed features, making them easier to model. This is especially helpful for linear models like Logistic Regression.
Features to Transform:
pdays: Apply a log transformation to normalize its distribution.


In [16]:
import numpy as np

# Apply log1p to handle skewed features
X_train['pdays_log'] = np.log1p(X_train['pdays'])
X_test['pdays_log'] = np.log1p(X_test['pdays'])


Polynomial Features
Why: Introducing interaction terms between features or polynomial features helps linear models (like Logistic Regression) or SVM with a linear kernel capture non-linear relationships. This can improve model performance when relationships between features and the target are non-linear.
Features: Since the dataset contains both numeric and categorical features, we can create polynomial features for numeric variables, such as age, campaign, previous, and interaction terms.

In [17]:
from sklearn.preprocessing import PolynomialFeatures

poly = PolynomialFeatures(degree=2, include_bias=False)
X_train_poly = poly.fit_transform(X_train_scaled)
X_test_poly = poly.transform(X_test_scaled)


In [19]:
from sklearn.metrics import f1_score, roc_auc_score
# Handle missing values by filling NaN values with mean for numeric columns and mode for categorical columns
X_train_scaled.fillna(X_train_scaled.mean(), inplace=True)
X_test_scaled.fillna(X_test_scaled.mean(), inplace=True)

# Initialize the models again after filling missing values
models = {
    'Logistic Regression': LogisticRegression(max_iter=1000),
    'K-Nearest Neighbors': KNeighborsClassifier(),
    'Support Vector Machine': SVC(probability=True)  # For AUC-ROC
}

# Dictionary to store results
results = {'Model': [], 'Train Accuracy': [], 'Test Accuracy': [], 'F1-Score': [], 'AUC-ROC': []}

# Train and evaluate each model
for model_name, model in models.items():
    # Train the model
    model.fit(X_train_scaled, y_train)
    
    # Predict on the test set
    y_test_pred = model.predict(X_test_scaled)
    y_test_prob = model.predict_proba(X_test_scaled)[:, 1] if hasattr(model, 'predict_proba') else None
    
    # Calculate metrics
    train_accuracy = model.score(X_train_scaled, y_train)
    test_accuracy = model.score(X_test_scaled, y_test)
    f1 = f1_score(y_test, y_test_pred)
    auc_roc = roc_auc_score(y_test, y_test_prob) if y_test_prob is not None else None
    
    # Store results
    results['Model'].append(model_name)
    results['Train Accuracy'].append(train_accuracy)
    results['Test Accuracy'].append(test_accuracy)
    results['F1-Score'].append(f1)
    results['AUC-ROC'].append(auc_roc)

# Convert results into a DataFrame
results_df = pd.DataFrame(results)



# Display the results
print(results_df)


                    Model  Train Accuracy  Test Accuracy  F1-Score   AUC-ROC
0     Logistic Regression        0.901335       0.897062  0.319422  0.780002
1     K-Nearest Neighbors        0.912504       0.889900  0.368824  0.714865
2  Support Vector Machine        0.903672       0.897427  0.319098  0.699183


Given the low F1 score, going to try SMOTE to handle imbalanced data

In [21]:
from imblearn.over_sampling import SMOTE
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, accuracy_score

# Step 1: Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 2: Apply standard scaling
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Step 3: Apply SMOTE to the training data
smote = SMOTE(random_state=42)
X_train_smote, y_train_smote = smote.fit_resample(X_train_scaled, y_train)

# Step 4: Train the model (e.g., Logistic Regression)
model = LogisticRegression(max_iter=1000)
model.fit(X_train_smote, y_train_smote)

# Step 5: Evaluate the model
y_pred = model.predict(X_test_scaled)
f1 = f1_score(y_test, y_pred)
accuracy = accuracy_score(y_test, y_pred)

print(f"F1-Score: {f1}")
print(f"Accuracy: {accuracy}")


F1-Score: 0.4314759036144578
Accuracy: 0.816703083272639


##### Questions